lib/medtronic/processData.js

```diff
@@ -417,23 +425,231 @@ function buildBGRecords(records) {
+  var postrecords = [];
+
+  suspendResumeRecords.forEach(function (entry) {
+    console.log('suspend/resume:',entry.jsDate, common.bytes2hex(entry.head), common.bytes2hex(entry.date), common.bytes2hex(entry.body));
+  });
+
+  for(var i=0; i < suspendResumeRecords.length; i++) {
+    var suspendEntry = suspendResumeRecords[i];
+
+    if(suspendEntry.type.value === RECORD_TYPES.PUMP_SUSPEND.value) {
+
+      var suspendedBasal = cfg.builder.makeSuspendBasal()
+        .with_deviceTime(sundial.formatDeviceTime(suspendEntry.jsDate))
+        .set('index', suspendEntry.index);
+      cfg.tzoUtil.fillInUTCInfo(suspendedBasal, suspendEntry.jsDate);
+      postrecords.push(suspendedBasal);
+
+      var suspendResume = cfg.builder.makeDeviceEventSuspendResume()
+        .with_deviceTime(sundial.formatDeviceTime(suspendEntry.jsDate))
+        .with_reason({suspended: 'manual', resumed: 'manual'}) // TODO: get reason
+        .set('index', suspendEntry.index);
+      cfg.tzoUtil.fillInUTCInfo(suspendResume, suspendEntry.jsDate);
+
+      var resumeEntry = suspendResumeRecords[i+1];
+      if(resumeEntry && resumeEntry.type.value === RECORD_TYPES.PUMP_RESUME.value) {
+        i += 1; // increment counter, as we have a matching resume
+        var duration = resumeEntry.jsDate.valueOf() - suspendEntry.jsDate.valueOf();
+        suspendResume.with_duration(duration);
+      } else {
+        annotate.annotateEvent(suspendResume,'status/incomplete-tuple');
+      }
+
+      postrecords.push(suspendResume.done());
+
+    } else {
+      throw new Error('Suspend/resume events out of order');
+    }
+
+  };
+  return postrecords;
+}
+
+function buildBasalProfiles(records) {
+
+  var getSchedule = function(record) {
+    var schedules = [];
+
+    if (record.head[1] === 0) {
+      return schedules; // record not set yet
+    }
+
+    for(var j = 0; j < 47; j++ ) {
+      var schedule = struct.unpack(record.body,j*3,'bbb',['offset', 'rate', 'q']);
+      if (schedule.offset === 0 && schedule.rate === 0 && schedule.q === 0) {
+        break;
+      }
+      var startTime = schedule.offset * 30 * sundial.MIN_TO_MSEC;
+      schedules.push( { start: startTime, rate: schedule.rate / cfg.settings.strokesPerUnit} );
+    };
+    return schedules;
+  };
+
+  // we start with the current basal profiles and work our way back
+  var basalProfiles = filterHistory([RECORD_TYPES.CHANGE_BASAL_PROFILE_OLD,
+                                 RECORD_TYPES.CHANGE_BASAL_PROFILE_NEW,
+                                 RECORD_TYPES.SELECT_BASAL_PROFILE], records).reverse();
+  var postrecords = [];
+  var scheduleChanges = [];
+  var activeSchedule, prevRecord;
+
+  // when we record schedule changes, we need to pre-fill with the
+  // most recent schedules, and remember that they only change for the
+  // *previous* schedules
+  var scheduleChange = _.clone(cfg.settings.basalSchedules);
+  scheduleChanges.push(scheduleChange,scheduleChange);
```

Comment:
I agree, it's very unintuitive. I think QA would be the best way to make sure it works as expected, as we'll be able to compare against CareLink driver.